### PR TITLE
Adding serialization support for std::variant (if available) and std::tuple

### LINKED ISSUE #4416: Add serialization support for std::variant and std::tuple
Feature request for completion and I actually need it:
Please add official HPX serialization support for std::variant and std::tuple

### PR DESCRIPTION
**Pull request by hkaiser**
- adding missing tests for tuple serialization

Fixes #4416

@McKillroy please verify whether this fixes the issue you reported

### CODE REVIEW

File: libs/serialization/tests/unit/serialization_boost_variant.cpp [hunk 1]
```diff
@@ -93,6 +102,7 @@ int main()
 
     HPX_TEST_EQ(sivar.which(), sovar.which());
     HPX_TEST_EQ(sivar, sovar);
+#endif
 
     return hpx::util::report_errors();
```
Comment (msimberg) [inline libs/serialization/tests/unit/serialization_boost_variant.cpp]: `#include <hpx/testing.hpp>` needs to be outside the `#ifdef` at the top for this, or this should just be `return 0` if `HPX_SERIALIZATION_HAVE_BOOST_TYPES` isn't defined (https://cdash.cscs.ch//viewBuildError.php?buildid=98967).

File: libs/serialization/tests/unit/serialization_tuple.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
+//  Copyright (c) 2020 Hartmut Kaiser
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/datatypes/tuple.hpp>
```
Comment (msimberg) [inline libs/serialization/tests/unit/serialization_tuple.cpp]:
```suggestion
#include <hpx/datastructures/tuple.hpp>
```

**Comment by msimberg** [comment]
Feel free to add more tests to the blacklists on the GH actions builders if it seems like they fail often.

**Comment by msimberg** [comment]
Is this ready to go or still in progress?

**Comment by hkaiser** [comment]
> Is this ready to go or still in progress?

This is ready to go from my end. Not sure if @biddisco is happy with how I have addressed his review comments.

**Comment by McKillroy** [comment]
Would be great - I'm just about to really need tuples and variants for my
messaging system. Can't wait for this to be in.  :)

**Comment by msimberg** [comment]
> > Is this ready to go or still in progress?
> 
> This is ready to go from my end. Not sure if @biddisco is happy with how I have addressed his review comments.

Looks good to me. I think @biddisco can let you know afterwards if he'd like to see more tests.

**Pull request merged**